cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -357,13 +355,20 @@ static inline Result<std::vector<RowGroupInfo>> AugmentRowGroups(
   return row_groups;
 }
 
-Result<ScanTaskIterator> ParquetFileFormat::ScanFile(
-    const FileSource& source, std::shared_ptr<ScanOptions> options,
-    std::shared_ptr<ScanContext> context, std::vector<RowGroupInfo> row_groups) const {
+Result<ScanTaskIterator> ParquetFileFormat::ScanFile(std::shared_ptr<ScanOptions> options,
+                                                     std::shared_ptr<ScanContext> context,
+                                                     FileFragment* fragment) const {
+  const auto& source = fragment->source();
+  auto row_groups = checked_cast<const ParquetFileFragment*>(fragment)->row_groups();
+
   bool row_groups_are_complete = RowGroupInfosAreComplete(row_groups);
   // The following block is required to avoid any IO if all RowGroups are
   // excluded due to prior statistics knowledge.
   if (row_groups_are_complete) {
+    // physical_schema should be cached at this point
+    ARROW_ASSIGN_OR_RAISE(auto physical_schema, fragment->ReadPhysicalSchema());
+    RETURN_NOT_OK(options->filter->Validate(*physical_schema));
```

Comment:
Didn't look in detail yet (so might be misreading, it's late here ;)), but a quick note: in general, when Fragments were instantiated with ParquetFactory (from a `_metadata` file) with populated statistics, I think it should be possible to filter out fragments without any additional IO.

Reply:
Correct, this line can't be here.

Reply:
Fragments created using `_metadata` have their physical schema preloaded so they won't incur IO when no row group can satisfy a predicate

Reply:
https://github.com/apache/arrow/pull/7526/files#diff-b7aed3b5c0adacff778c6930cc6a03e4R646-R650

Reply:
To avoid confusion about this, should we rename ReadPhysicalSchema? Or instead of having `ReadPhysicalSchema` which caches and `ReadPhysicalSchemaImpl` that always reads, a `GetPhysicalSchema` that caches and `ReadPhysicalSchema` that reads?